Mathlib/CategoryTheory/ChosenFiniteProducts.lean

```diff
@@ -228,6 +228,271 @@ instance (priority := 100) : Limits.HasFiniteProducts C :=
+      whiskerRight (prodComparisonBifunctorNatTrans F) ((whiskeringRight _ _ _).obj G) ≫
+        whiskerLeft F (whiskerRight (prodComparisonBifunctorNatTrans G)
+          ((whiskeringLeft _ _ _).obj F)) := by ext; simp [prodComparison_comp]
+
+instance (A : C) [∀ B, IsIso (prodComparison F A B)] : IsIso (prodComparisonNatTrans F A) := by
+  letI : ∀ X, IsIso ((prodComparisonNatTrans F A).app X) := by assumption
+  apply NatIso.isIso_of_isIso_app
+
+instance [∀ A B, IsIso (prodComparison F A B)] : IsIso (prodComparisonBifunctorNatTrans F) := by
+  letI : ∀ X, IsIso ((prodComparisonBifunctorNatTrans F).app X) :=
+    fun _ ↦ by dsimp; apply NatIso.isIso_of_isIso_app
+  apply NatIso.isIso_of_isIso_app
+
+open Limits
+section PreservesLimitPairs
+
+section
+variable (A B)
+variable [PreservesLimit (pair A B) F]
+
+/-- If `F` preserves the limit of the pair `(A, B)`, then the binary fan given by
+`(F.map fst A B, F.map (snd A B))` is a limit cone. -/
+def isLimitChosenFiniteProductsOfPreservesLimits :
+    IsLimit <| BinaryFan.mk (F.map (fst A B)) (F.map (snd A B)) :=
+  mapIsLimitOfPreservesOfIsLimit F (fst _ _) (snd _ _) <|
+    (product A B).isLimit.ofIsoLimit <| isoBinaryFanMk (product A B).cone
+
+/-- If `F` preserves the limit of the pair `(A, B)`, then `prodComparison F A B` is an isomorphism.
+-/
+def prodComparisonIso : F.obj (A ⊗ B) ≅ F.obj A ⊗ F.obj B :=
+  IsLimit.conePointUniqueUpToIso (isLimitChosenFiniteProductsOfPreservesLimits F A B)
+    (product _ _).isLimit
+
+@[simp]
+lemma prodComparisonIso_hom : (prodComparisonIso F A B).hom = prodComparison F A B := by
+  rfl
+
+instance isIso_prodComparison_of_preservesLimit_pair: IsIso (prodComparison F A B) := by
+  rw [← prodComparisonIso_hom]
+  infer_instance
+
+end
+
+/-- The natural isomorphism `F(A ⊗ -) ≅ FA ⊗ F-`, provided each `prodComparison F A B` is an
+isomorphism (as `B` changes). -/
+@[simps! hom inv]
+noncomputable def prodComparisonNatIso (A : C) [∀ B, PreservesLimit (pair A B) F] :
+    (curriedTensor C).obj A ⋙ F ≅ F ⋙ (curriedTensor D).obj (F.obj A) :=
+  asIso (prodComparisonNatTrans F A)
+
+/-- The natural isomorphism of bifunctors `F(- ⊗ -) ≅ F- ⊗ F-`, provided each
+`prodComparison F A B` is an isomorphism. -/
+@[simps! hom inv]
+noncomputable def prodComparisonBifunctorsNatIso [∀ A B, PreservesLimit (pair A B) F] :
+    curriedTensor C ⋙ (whiskeringRight _ _ _).obj F ≅
+      F ⋙ curriedTensor D ⋙ (whiskeringLeft _ _ _).obj F :=
+  asIso (prodComparisonBifunctorNatTrans F)
+
+end PreservesLimitPairs
+
+section ProdComparisonIso
+
+/-- If `prodComparison F A B` is an isomorphism, then `F` preserves the limit of `pair A B`. -/
+noncomputable def preservesLimitPairOfIsIsoProdComparison (A B : C) [IsIso (prodComparison F A B)] :
+    PreservesLimit (pair A B) F := by
+  apply preservesLimitOfPreservesLimitCone (product A B).isLimit
+  apply IsLimit.equivOfNatIsoOfIso (pairComp A B F) _
+    ((product (F.obj A) (F.obj B)).cone.extend (prodComparison F A B)) _|>.invFun
```

Comment:
Could you add the following definition in `Limits.Shapes.BinaryProducts`:
```lean
def BinaryFan.ext {A B : C} {c c' : BinaryFan A B} (e : c.pt ≅ c'.pt)
    (h₁ : e.hom ≫ c'.fst = c.fst) (h₂ : e.hom ≫ c'.snd = c.snd) : c ≅ c' :=
  sorry
```
as a particular case of more general missing `Fan.ext` which you should add in `Limits.Shapes.Products`.
Then, we can get a slightly better code here:
```suggestion
  apply preservesLimitOfPreservesLimitCone (product A B).isLimit
  refine IsLimit.equivOfNatIsoOfIso (pairComp A B F) _
    ((product (F.obj A) (F.obj B)).cone.extend (prodComparison F A B))
      (BinaryFan.ext (by exact Iso.refl _) ?_ ?_) |>.invFun
      (IsLimit.extendIso _ (product (F.obj A) (F.obj B)).isLimit)
```